### PR TITLE
Blogpost facet navigation

### PR DESCRIPTION
**Pull request by dieterve**
I need some input on the title, I struggled with it. I don't think its clear enough that its about facets plus its counters.

Input about my style is also welcome.

### CODE REVIEW

**Comment by dieterve** [comment]
Not sure about my stylish image as well. Maybe somebody has better suggestion?

File: _posts/2016-03-14-faceted-navigation.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
+---
+layout: post
+title: "Howto build faceted search and facet counters"
```
Comment (hannesvdvreken) [inline _posts/2016-03-14-faceted-navigation.md]: `How to` with a space, no? or was this intentional?

File: _posts/2016-03-14-faceted-navigation.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+date: 2016-03-14
+author: dieterve
+categories: ["elasticsearch", "search"]
+
+---
+
+We all know them, the search pages which allow you to filter through vast data sets by checking or unchecking filters. In most cases each filter is followed by a counter which indicates how many results will be shown when you use that filter. Those counters inform the user about his next move before he performs it.
```
Comment (hannesvdvreken) [inline _posts/2016-03-14-faceted-navigation.md]: before he performs it. Personally I would make it `he/she`.

File: _posts/2016-03-14-faceted-navigation.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+### Requirements
+
+![Faceted search without selected filters](/assets/posts/2016-03-14-faceted-navigation/default-facets.png)
+
+Before we can build anything, we need to understand how we expect faceted search to behave. This might be different depending on your business case. The best way to explain this is to use an example. Lets assume we have a search with 2 facets: category and country. Within one facet, the filters act as an `OR`. Across the facets, it acts as an `AND`. In pseudo query this would look like `WHERE (category = 'books' OR category = 'movies') AND (country = 'BE' OR country = 'FR')`. This would get all the books and movies in Belgium and France.
+
+The counters need to take the current active filters in to account and add themselves. So if our current active search filter is `WHERE category = 'books' AND country = 'BE'`. The query to get the count for the France filter would be `COUNT WHERE category = 'books' AND (country = 'BE' OR country = 'FR')`. This count needs to be executed on the entire dataset, not on the filtered dataset. If you count on the filtered dataset you will bias your counters and they'll actually be lower then expected.
```
Comment (hannesvdvreken) [inline _posts/2016-03-14-faceted-navigation.md]: in to -> into

Comment (hannesvdvreken) [inline _posts/2016-03-14-faceted-navigation.md]: > The counters need to take the current active filters in to account and add themselves.

I don't get this

File: _posts/2016-03-14-faceted-navigation.md (hunk 2 above)
Comment (miggynascimento) [inline _posts/2016-03-14-faceted-navigation.md]: `they` works as well

File: _posts/2016-03-14-faceted-navigation.md (hunk 3 above)
Comment (hannesvdvreken) [inline _posts/2016-03-14-faceted-navigation.md]: lower then expected -> lower than expected

File: _posts/2016-03-14-faceted-navigation.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+          },
+          {
+              "terms": {"category": ["books", "movies"]}
+          }
+      ]
+   }
+}
+```
+
+For the counters we can use the build in [aggregations](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations.html) from elasticsearch. Each of our 2 facets are stored as a single field in an index, so we can use a [terms](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html) aggregation on each of those fields. The aggregation will return a counter per value of that field. Exactly what we need.
```
Comment (hannesvdvreken) [inline _posts/2016-03-14-faceted-navigation.md]: build in -> built-in

File: _posts/2016-03-14-faceted-navigation.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+From a technical standpoint, there are a few challenges to those counters. To know the exact count we have to execute the search query again but with that filter. In a complex search application this quickly amounts to a large number of queries. Luckily Elasticsearch, in all its glory, allows us to do this in one query.
+
+### Requirements
+
+![Faceted search without selected filters](/assets/posts/2016-03-14-faceted-navigation/default-facets.png)
+
+Before we can build anything, we need to understand how we expect faceted search to behave. This might be different depending on your business case. The best way to explain this is to use an example. Lets assume we have a search with 2 facets: category and country. Within one facet, the filters act as an `OR`. Across the facets, it acts as an `AND`. In pseudo query this would look like `WHERE (category = 'books' OR category = 'movies') AND (country = 'BE' OR country = 'FR')`. This would get all the books and movies in Belgium and France.
```
Comment (miggynascimento) [inline _posts/2016-03-14-faceted-navigation.md]: `Let's`

File: _posts/2016-03-14-faceted-navigation.md (hunk 4 above)
Comment (hannesvdvreken) [inline _posts/2016-03-14-faceted-navigation.md]: from elasticsearch. -> from **E**lasticsearch.

File: _posts/2016-03-14-faceted-navigation.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+                  }
+               }
+            }
+         }
+      }
+   }
+}
+```
+
+This query finally produces the results and the counters we were after. This is a simple query with only 2 filters, when you have a large set of filters and different types of filters it is easy to lose overview and allow bugs to sneak in. You might want to look into a package like our [elasticsearcher](https://github.com/madewithlove/elasticsearcher) to help you abstract parts of your query so you have a better overview.
```
Comment (hannesvdvreken) [inline _posts/2016-03-14-faceted-navigation.md]: > This is a simple query with only 2 filters, when you have a large set

This is a simple query with only 2 filters **but** when you have a large set

Comment (hannesvdvreken) [inline _posts/2016-03-14-faceted-navigation.md]: it is easy to lose **your** overview..?

**Comment by hannesvdvreken** [comment]
Nice post!

I would also include an example JSON output of what ES returns for that query.

What if you select "books" and you need to recalculate the counts for other filters? Maybe an idea for a follow up article.

**Comment by hannesvdvreken** [comment]
You like this image? screen shot 2016-03-21 at 21 29 55

**Comment by tonysm** [comment]
Really good one! :clap:

**Comment by dieterve** [comment]
Tx for the feedback, I did some cleaning up.

@hannesvdvreken: tx for the image, included it. Can you check if it is still unclear.

Not sure about including the responses, they won't make a lot of sense without knowing/having the dataset. It would for example look like this:

```json
{
   "took": 153,
   "timed_out": false,
   "_shards": {
      "total": 5,
      "successful": 5,
      "failed": 0
   },
   "hits": {
      "total": 3,
      "max_score": 0,
      "hits": []
   },
   "aggregations": {
      "all_products": {
         "doc_count": 21,
         "filterted categories": {
            "doc_count": 13,
            "categories": {
               "doc_count_error_upper_bound": 0,
               "sum_other_doc_count": 0,
               "buckets": [
                  {
                     "key": "movies",
                     "doc_count": 6
                  },
                  {
                     "key": "music",
                     "doc_count": 4
                  },
                  {
                     "key": "books",
                     "doc_count": 3
                  }
               ]
            }
         },
         "filtered_countries": {
            "doc_count": 15,
            "countries": {
               "doc_count_error_upper_bound": 0,
               "sum_other_doc_count": 0,
               "buckets": [
                  {
                     "key": "fr",
                     "doc_count": 6
                  },
                  {
                     "key": "br",
                     "doc_count": 4
                  },
                  {
                     "key": "be",
                     "doc_count": 3
                  },
                  {
                     "key": "pt",
                     "doc_count": 2
                  }
               ]
            }
         }
      }
   }
}
```

It is a lot of data which I think isn't very useful... What do you think?

**Comment by hannesvdvreken** [comment]
You could shrink the response by using `...` and/or put it in a gist link or something.

**Comment by dieterve** [comment]
Oke, good to go for me!

**Pull request merged**